src/algorithms/mts.jl

```diff
@@ -0,0 +1,354 @@
+        update_xk = true
+        # Heuristic search
+        _xk_x1, _xk_y1, _xk_x2  = copy(_xk), copy(_xk), copy(_xk)
+        _xk_x1[i] += 0.1
+        _xk_y1[i] -= 0.1
+        _xk_x2[i] += 0.2
+        _xk_x1_val, _xk_y1_val, _xk_x2_val = evaluate!(model, _xk_x1), evaluate!(model, _xk_y1), evaluate!(model, _xk_x2)
+        if _xk_x1_val < workspace.optimal_val
+            grade += BONUS1
+            workspace.optimal_x, workspace.optimal_ind, workspace.optimal_val = copy(_xk), k, _xk_val
+        end
+        if _xk_y1_val < workspace.optimal_val
+            grade += BONUS1
+            workspace.optimal_x, workspace.optimal_ind, workspace.optimal_val = copy(_xk), k, _xk_val
+        end
+        if _xk_x2_val < workspace.optimal_val
+            grade += BONUS1
+            workspace.optimal_x, workspace.optimal_ind, workspace.optimal_val = copy(_xk), k, _xk_val
+        end
+        D1, D2, D3 = _xk_val - _xk_x1_val, _xk_val - _xk_y1_val, _xk_val - _xk_x2_val
+        grade += ((D1>0) + (D2>0) + (D3>0)) * BONUS2
+        if update_xk == false
+            update_xk = (D1>0) || (D2>0) || (D3>0)
+        end
+        a = (rand() * (a_max - a_min)) + a_min
+        b = (rand() * (b_max - b_min)) + b_min
+        c = (rand() * (c_max - c_max)) + c_min
+        _xk[i] += a*(D1 - D2) + b*(D3 - 2*D1) + c
+    end
+    if update_xk
+        # Update x[k]
+        x[k] = _xk
+    end
+    return grade
+end
+
+# Optimization using localsearch1
+function localsearch1(workspace::MTSWorkspace)
+    M = workspace.options.M
+    for i in 1:M
+        _localsearch1(workspace, i)
+    end
+end
+
+# Optimization using localsearch2
+function localsearch2(workspace::MTSWorkspace)
+    M = workspace.options.M
+    for i in 1:M
+        _localsearch2(workspace, i)
+    end
+end
+
+# Optimization using localsearch3
+function localsearch3(workspace::MTSWorkspace)
+    M = workspace.options.M
+    for i in 1:M
+        _localsearch3(workspace, i)
+    end
+end
+
+const LOCAL_SEARCH_METHODS = [_localsearch1, _localsearch2, _localsearch3]
+const N_METHODS = length(LOCAL_SEARCH_METHODS)
+
+function MTS(workspace::MTSWorkspace)
+    # Multiple Trajectory Search
+    @unpack options, enable = workspace
+    @unpack M, n_foreground, n_local_search, n_local_search_test, n_local_search_best = options
+    grade_x = [-Inf for _ in 1:M]
+    for i in 1:M
+        if !enable[i]
+            continue
+        end
+        LS_testgrades = [0 for _ in 1:N_METHODS]
+        for _ in 1:n_local_search_test
+            LS_testgrades = [LS_testgrades[m]+_local_search(workspace, i) for (m, _local_search) in enumerate(LOCAL_SEARCH_METHODS)]
+        end
+        _best_local_search = LOCAL_SEARCH_METHODS[argmax(LS_testgrades)]
+        grade_x[i] = 0
+        for _ in 1:n_local_search
+            grade_x[i] += _best_local_search(workspace, i)
+        end
+    end
+    for _ in 1:n_local_search_best
+        _localsearch1(workspace, workspace.optimal_ind)
```

Comment:
shouldn't you use the best one here instead of the first?

Reply:
No they always use ls1 at the end of every iteration, that's the reason why ls1 is a "main algorithm" among 3 subalgorithms. Also explained why people can use ls1 independently.